File: go/internal/approval/scope.go
/**
 * Author: Deepankar Das
 */

package approval

import (
	"strings"
	"time"

	"github.com/anthropics/enforcer/internal/types"
)

// MatchesScope checks whether an incoming action request matches the given
// approval scope.
//
// Scope semantics:
//   - Single-use scopes (ScopeSingle) NEVER match here; they are consumed on
//     first use and removed by the caller (ApprovalService.CheckScope).
//     Returning false ensures that a single-use scope is never reused.
//   - Time-bounded scopes (ScopeTimeBounded) match only if the current time
//     is before the scope's expiry.
//   - Session scopes (ScopeSession) match as long as the pattern matches.
//
// Pattern matching:
//   - An empty pattern matches everything.
//   - The pattern is compared against the action type (string prefix) and
//     the resource value (wildcard prefix: "foo.*" matches "foo.bar").
func MatchesScope(request types.ActionRequest, scope types.ApprovalScope) bool {
	// Single-use scopes match once and are consumed by CheckScope after use.

	// Time-bounded scopes check expiry.
	if scope.Type == types.ScopeTimeBounded {
		if scope.Expiry != "" {
			expiry, err := time.Parse(time.RFC3339, scope.Expiry)
			if err != nil {
				return false
			}
			if time.Now().UTC().After(expiry) {
				return false
			}
		}
	}

	// Pattern matching.
	if scope.Pattern == "" {
		return true
	}

	actionType := string(request.Action.Type)
	resourceValue := request.Resource.Value
	if resourceValue == "" {
		resourceValue = request.Resource.Path
	}
	if resourceValue == "" {
		resourceValue = request.Resource.Host
	}

	// Exact match on action type prefix.
	if strings.HasPrefix(actionType, scope.Pattern) {
		return true
	}

	// Wildcard prefix match on resource value: "foo.*" matches "foo.bar".
	if strings.HasSuffix(scope.Pattern, ".*") {
		prefix := scope.Pattern[:len(scope.Pattern)-2]
		if strings.HasPrefix(actionType, prefix) {
			return true
		}
		if strings.HasPrefix(resourceValue, prefix) {
			return true
		}
	}

	// Direct resource value match.
	if resourceValue != "" && strings.HasPrefix(resourceValue, scope.Pattern) {
		return true
	}

	return false
}

File: src/approval/scope.ts
/**
 * Author: Deepankar Das
 */

/**
 * Enforcer — Reusable Approval Scopes
 *
 * Checks if a new action matches an existing approved scope.
 * When a scope matches, the action is auto-approved without re-prompting.
 *
 * Scope types:
 *   - single: one action only (default — no reuse)
 *   - session: all matching actions in this session
 *   - time_bounded: matching actions until expiry
 *
 * TDD Reference: Section 10.2
 * PRD Reference: Appendix C, R-4
 */

import type { ActionRequest } from "../../types/action.js";
import type { ApprovalScope } from "../../types/approval.js";

/**
 * Check if an action matches an approval scope.
 *
 * Pattern matching:
 *   - Exact match: "npm install express" matches "npm install express"
 *   - Wildcard prefix: "npm install *" matches any "npm install ..."
 *   - Action type prefix: "shell.exec:*" matches any shell command
 */
export function matchesScope(
  request: ActionRequest,
  scope: ApprovalScope,
): boolean {
  // Single-use scopes never match a second time
  if (scope.type === "single") {
    return false;
  }

  // Time-bounded scopes: check expiry
  if (scope.type === "time_bounded" && scope.expiry) {
    if (new Date(scope.expiry).getTime() < Date.now()) {
      return false; // Expired
    }
  }

  // If no pattern, scope matches everything in the session
  if (!scope.pattern) {
    return true;
  }

  const pattern = scope.pattern;
  const actionString = `${request.action.type}:${request.resource.value || request.resource.path || ""}`;

  // Exact match
  if (actionString === pattern) {
    return true;
  }

  // Wildcard matching
  if (pattern.endsWith("*")) {
    const prefix = pattern.slice(0, -1);
    if (actionString.startsWith(prefix)) {
      return true;
    }
    // Also check just the resource value
    const resourceValue = request.resource.value || request.resource.path || "";
    if (resourceValue.startsWith(prefix)) {
      return true;
    }
  }

  // Check resource value directly against pattern
  const resourceValue = request.resource.value || request.resource.path || "";
  if (resourceValue === pattern) {
    return true;
  }

  return false;
}